cpp/src/utilities/cudf_utils.h

```diff
@@ -27,4 +29,65 @@ gdf_size_type gdf_get_num_chars_bitmask(gdf_size_type size) {
 	return (( size + ( GDF_VALID_BITSIZE - 1)) / GDF_VALID_BITSIZE ); 
 }
 
-#endif
+/* --------------------------------------------------------------------------*/
+/** 
+ * @brief Flatten AOS info from gdf_columns into SOA.
+ * 
+ * @Param[in] cols Host-side array of gdf_columns
+ * @Param[in] ncols # columns
+ * @Param[out] d_cols Pointer to device array of columns
+ * @Param[out] d_types Device array of column types
+ * 
+ * @Returns
+ */
+/* ----------------------------------------------------------------------------*/
+inline void soa_col_info(gdf_column* cols, size_t ncols, void** d_cols, int* d_types)
+{
+	std::vector<void*> v_cols(ncols, nullptr);
+	std::vector<int>   v_types(ncols, 0);
+	for(size_t i=0; i<ncols; ++i)
+	{
+		v_cols[i] = cols[i].data;
+		v_types[i] = cols[i].dtype;
+	}
+
+	void** h_cols = &v_cols[0];
+	int* h_types = &v_types[0];
+	cudaMemcpy(d_cols, h_cols, ncols*sizeof(void*), cudaMemcpyHostToDevice);//TODO: add streams
```

Comment:
Wrap in `CUDA_TRY`. https://github.com/rapidsai/cudf/blob/master/cpp/src/utilities/error_utils.h#L8

You'll need to update `soa_col_info` to return a `gdf_error` and check the result at the call site.